packages/scheduler/lib/models/tasks.integration.test.ts

```diff
@@ -118,6 +119,62 @@ describe('Task', () => {
         const dequeued = (await tasks.dequeue(db, { groupKey: task.groupKey, limit: 1 })).unwrap();
         expect(dequeued).toHaveLength(0);
     });
+    it('should be dequeued according to group max concurrency ', async () => {
+        const groupKey = 'A';
+        const groupMaxConcurrency = 2;
+        const t0 = await createTask(db, { groupKey, groupMaxConcurrency });
+        const t1 = await createTask(db, { groupKey, groupMaxConcurrency });
+
+        let dequeued = (await tasks.dequeue(db, { groupKey, limit: 10 })).unwrap();
+        expect(dequeued).toHaveLength(2);
+        expect(dequeued[0]).toMatchObject({ id: t0.id, state: 'STARTED' });
+        expect(dequeued[1]).toMatchObject({ id: t1.id, state: 'STARTED' });
+
+        // group has reached its max concurrency, so no more tasks should be dequeued
+        const t2 = await createTask(db, { groupKey, groupMaxConcurrency });
+        dequeued = (await tasks.dequeue(db, { groupKey, limit: 10 })).unwrap();
+        expect(dequeued).toHaveLength(0);
+
+        // dequeuing tasks with different group key should not be affected
+        const t3 = await createTask(db, { groupKey: 'B', groupMaxConcurrency });
+        dequeued = (await tasks.dequeue(db, { groupKey: 'B', limit: 10 })).unwrap();
+        expect(dequeued).toHaveLength(1);
+        expect(dequeued[0]).toMatchObject({ id: t3.id, state: 'STARTED' });
+
+        // tasks now completes
+        await succeedTask(db, t0.id);
+        await succeedTask(db, t1.id);
+
+        // group should be able to dequeue again
+        dequeued = (await tasks.dequeue(db, { groupKey, limit: 10 })).unwrap();
+        expect(dequeued).toHaveLength(1);
+        expect(dequeued[0]).toMatchObject({ id: t2.id, state: 'STARTED' });
+    });
+    it('should respect group max concurrency with parallel dequeue calls', async () => {
+        const groupKey = 'A';
+        const groupMaxConcurrency = 2;
+
+        // creating and dequeing tasks in parallel in a tight loop
+        // to increase the chance of race conditions
```

Comment:
not the cleanest code below but simulating parallelism with nodejs is tricky. Let me know if you known a better way